capa/features/extractors/ghidra/file.py

```diff
@@ -0,0 +1,198 @@
+       matching:
+     - modulename.importname
+     - importname
+    """
+
+    fm = currentProgram.getFunctionManager().getExternalFunctions()
+    while fm.hasNext():
+        f = fm.next()
+        addr = int(f.getEntryPoint().toString().split(':')[1], 16)  # format: EXTERNAL:<hex_addr>
+        fstr = f.getName()  # format: 'importname' / 'Ordinal_*'
+        if 'Ordinal_' in fstr:
+            fstr = f"#{fstr.split('_')[1]}"
+        yield Import(fstr), AbsoluteVirtualAddress(addr)
+
+
+def extract_file_section_names() -> Iterator[Tuple[Feature, Address]]:
+    """extract section names"""
+
+    for block in currentProgram.getMemory().getBlocks():
+        yield Section(block.getName()), AbsoluteVirtualAddress(block.getStart().getOffset())
+
+
+def extract_file_strings() -> Iterator[Tuple[Feature, Address]]:
+    """extract ASCII and UTF-16 LE strings"""
+
+    dat = getFirstData()
+    while (dat != None):
+        if (dat.hasStringValue()):
+            yield String(dat.getValue()), FileOffsetAddress(dat.getAddress().getOffset()) 
+        dat = getDataAfter(dat)
+
+
+def extract_file_function_names() -> Iterator[Tuple[Feature, Address]]:
+    """
+    extract the names of statically-linked library functions.
+    """
+
+    fm = currentProgram.getFunctionManager().getExternalFunctions()
+    while fm.hasNext():
+        f = fm.next()
+        addr = int(f.getEntryPoint().toString().split(':')[1], 16)  # format: EXTERNAL:<hex_addr>
+        name = f.getName()
+        yield FunctionName(name), AbsoluteVirtualAddress(addr)
+        if name.startswith("_"):
+            # some linkers may prefix linked routines with a `_` to avoid name collisions.
+            # extract features for both the mangled and un-mangled representations.
+            # e.g. `_fwrite` -> `fwrite`
+            # see: https://stackoverflow.com/a/2628384/87207
+            yield FunctionName(name[1:]), AbsoluteVirtualAddress(addr)
+
+
+def extract_file_format() -> Iterator[Tuple[Feature, Address]]:
+
+    ef = currentProgram.getExecutableFormat()
+    if 'PE' in ef:
+        yield Format(FORMAT_PE), NO_ADDRESS
+    elif 'ELF' in ef:
+        yield Format(FORMAT_ELF), NO_ADDRESS
+    elif 'Raw' in ef:
+        # no file type to return when processing a binary file, but we want to continue processing
+        return
+    else:
+        raise NotImplementedError(f"unexpected file format: {view_type}")
+
+
+def extract_features() -> Iterator[Tuple[Feature, Address]]:
+    """extract file features"""
+    for file_handler in FILE_HANDLERS:
+        for feature, addr in file_handler():
+            yield feature, addr
+
+
+FILE_HANDLERS = (
+    extract_file_export_names,
+    extract_file_import_names,
+    extract_file_strings,
+    extract_file_section_names,
+    extract_file_embedded_pe,
+    extract_file_function_names,
+    extract_file_format,
+)
+
+
+def main():
+    """ """
+    import pprint
+
+    pprint.pprint(list(extract_features()))
```

Comment:
could be good to compare these against vivisect/IDA extracted features for potential differences

Reply:
Good call, thanks! I'll generate some data from the various backends and `diff` them to see where I can clean things up.